text/0193-removal-of-type-unions.md

```diff
@@ -0,0 +1,484 @@
+  be unable to determine which implementation to dynamically dispatch to.
+
+Additionally, certain static languages (such as **Go**) do not support overloading at all, rendering this approach
+entirely unsuitable.
+
+# Design Summary
+
+Removal of _type unions_ from the AWS CDK requires designing new APIs for those places where they are in use. This
+includes two kinds of APIs:
+
+- [Generated as part of the CloudFormation Resources](#generated-code)
+- [Manually authored as part of the AWS Construct Library or AWS CDK Core framework](#hand-written-code)
+
+# Detailed Design
+
+## Generated code
+
+As previously discussed, the _CloudFormation Resources_ ubiquitously rely on _type unions_ to allow providing late-bound
+values to the various properties accepted by the resources. Late-binding values is necessary to support mutation APIs
+which provide enhanced ergonomics for certain constructs, particularly where various components can be expected to
+collaborate on the finalized configuration of a resource.
+
+All _type unions_ where late-bound values can be represented in the literal type of the value (`string`, `number` and
+`any`) can be simplified to that literal type, since the required APIs already exist and the migration path for those
+use-cases is already very well defined: using `Lazy` factory methods to manually wrap `IResolvable` instances.
+
+On the other hand, _type unions_ of `IResolvable` and complex types (including collections) cannot be resolved that
+easily. In **TypeScript** and other dynamic languages (such as **Python**), the `Lazy.asAny` method can be used to wrap
+an `IResolvable` and evade the type checker. This approach is however no workable in **C#**, **Go** and **Java** (and
+likely other statically typed languages we may wish to support in the future).
+
+For those cases, one solution is to have users of statically typed languages use _[escape hatches]_ to forcefully
+override properties with the desired late-bound value, since _[escape hatches]_ offer `any`-typed accessors. This
+results in a degradation of ergonomics for statically typed languages, but not not imply a redution in available
+features offered by the AWS CDK.
+
+An option is to generate specific wrapper types that effectively encapsulate the _type union_, however such wrapper
+types must then be used for _all_ properties in order to avoid incurring an API breaking change if some _CloudFormation
+Resource_ schema adds support for alternate value types on a given property (this is possible since CloudFormation
+describes resource schemas using [JSON Schema]). This would result in a significant deterioration of the ergonomics of
+the _CloudFormation Resource_ classes, but provides an API that is safe to use from any programming language.
+
+[escape hatches]: https://docs.aws.amazon.com/cdk/latest/guide/cfn_layer.html
+[json schema]: https://json-schema.org
```

Comment:
I've read this several times, but I still don't understand what the proposed solution is here.

Can you show:

a) What the new generated code for Cfn* classes will look like in TypeScript.
b) What will the experience for the Cfn* classes look like for other languages (probably just Java and Python are representative enough to cover 99% of the cases).

Reply:
Feel free to use something like `CfnBucket` as a reference.

Reply:
Fair enough, will illustrate stuff a lot more.

The fact a bunch of places lists options and doesn't quite make recommendations was intentional at this point (I needed a base to discuss with several folks before I'd narrow it down). But I wholeheartedly agree it does not help understanding the proposal, so I'll revise.